### ISSUE TITLE
'a:f32[32,12,512,512] = add_any b c' cannot be delayed in apply_grad.py

### ISSUE THREAD

**Issue by cksmll** [alpa-projects/alpa#808]
**Please describe the bug**
I use the alpa for T5-base model and get some errors below.
I get the baseline code(run_t5_mlm_flax.py) in https://github.com/huggingface/transformers/blob/main/examples/flax/language-modeling/run_t5_mlm_flax.py
And revise the code similar to alpa/example/op_finetune/run_clm_flax.py
Overall training code and run_script is here: https://gist.github.com/cksmll/1663014f698d15c3fb6a665578ad7c99

**Screenshots**
Error:
<img width="1374" alt="image" src="https://user-images.githubusercontent.com/119835217/206180101-8f3b48a6-0016-4abf-891c-02cc6e44c752.png">
The error is in the v0.2.2

When I get the main branch in this github by pip install git+https://github.com/alpa-projects/alpa@main
I got the dictionary key error(asssertionError)
**Please describe the expected behavior**
I want to run the alpa's pipeshard class in T5
**System information and environment**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04, docker): Ubuntu 20.04
- Python version: 3.8
- CUDA version:11.3
- NCCL version:
- cupy version: cupy-cuda113
- GPU model and memory: RTX 3090
- 
<img width="566" alt="image" src="https://user-images.githubusercontent.com/119835217/206181662-ef77d1d5-0361-4678-b11e-baf433e5486b.png">

- Alpa version: v0.2.2
- TensorFlow version: 2.10.0
- JAX version:0.3.22

**To Reproduce**
1. This is my script: https://gist.github.com/cksmll/1663014f698d15c3fb6a665578ad7c99
2. To get the real data and tokenizer, you need some time but in this script, I run the fake_batch (jnp.ones) in training loop. So you could turn off the loading of data and tokenizer.

**Comment by cksmll** [comment]
Oh, After getting the new version It runs. (Until a few days ago, the latest code did not solve it.)

However, it only runs when the tensorflow is not installed.
If I run the script after pip3 install tensorflow, the following memory error occurs.
<img width="1248" alt="image" src="https://user-images.githubusercontent.com/119835217/206222955-8538d8a4-53d2-4549-a7a2-7faa7509103d.png">

I wonder why

**Comment by ZYHowell** [comment]
This seems an OOM error. If the model is not very large, maybe it is related to [this issue](https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html). For example, the XLA from jaxlib will try to get 90% of all available memory by default. If there is two XLA launched separately(alpa's jaxlib and tensorflow), one may only get 90% of only 10% of all memory, because another lib already occupies 90% of all memory

**Comment by cksmll** [comment]
Thank you for answering.
But again I got the dictionary keyError in Alpa@main version.
This is the error for T5-11b model. (huggingface: https://huggingface.co/t5-11b)
<img width="1152" alt="image" src="https://user-images.githubusercontent.com/119835217/206400930-7863415f-b83d-43aa-8bd9-d7a7116e9e42.png">
This error can't see in T5-base, T5-large.
As you advise, there is no tensorflow.

This is my script: https://gist.github.com/cksmll/1663014f698d15c3fb6a665578ad7c99

**Comment by ZYHowell** [comment]
Actually WIP with the bug: https://github.com/alpa-projects/alpa/pull/807

**Comment by ZYHowell** [comment]
@cksmll Could you please try the nightly alpa after https://github.com/alpa-projects/alpa/pull/807?

**Comment by cksmll** [comment]
Thanks. It finally works well with T5-11b!!!
Lastly, I want to ask you one minor thing. (Sorry for bothering you)

For training the T5, I have to turn off the drop_out layer (setting the 'Train=False' in `train_step` function).
Otherwise, if I turn on the drop_out layer like this. It prints error.
error:
<img width="654" alt="image" src="https://user-images.githubusercontent.com/119835217/206863682-cd6e30a5-6071-4651-b307-78978c90c1ca.png">
How I apply the drop_out layer:
<img width="621" alt="image" src="https://user-images.githubusercontent.com/119835217/206863758-a9bcc0be-b7dd-4beb-9884-d92bc0ac9440.png">

So, To use the alpa, Is it right to turn off the drop_out layer?

**Comment by ZYHowell** [comment]
It seems like another bug...I'll try to fix it.

**Comment by ZYHowell** [comment]
@cksmll  In alpa, we monkey patches the rng from jax's stateless version to tf's stateful one, and jax uses a specific dtype for its rng, which is not handled. As a walkaround, could you please try to move the key our of the `loss_fn`? (and imo, for the literal correctness, you need to move it out of `train_step` to make each iteration not dropout at the same position)
